pkg/datapath/tables/node_address.go

```diff
@@ -0,0 +1,365 @@
+	defer n.tracker.Close()
+
+	limiter := rate.NewLimiter(nodeAddressControllerMinInterval, 1)
+	revision := statedb.Revision(0)
+	for {
+		txn := n.DB.WriteTxn(n.NodeAddresses)
+		process := func(dev *Device, deleted bool, rev statedb.Revision) error {
+			addrIter, _ := n.NodeAddresses.Get(txn, NodeAddressDeviceNameIndex.Query(dev.Name))
+			existing := sets.New[NodeAddress](statedb.Collect[NodeAddress](addrIter)...)
+			var new sets.Set[NodeAddress]
+			if !deleted {
+				new = n.getAddressesFromDevice(dev)
+			}
+			n.update(txn, existing, new, reporter)
+			return nil
+		}
+		var watch <-chan struct{}
+		revision, watch, _ = n.tracker.Process(txn, revision, process)
+		txn.Commit()
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-watch:
+		}
+		if err := limiter.Wait(ctx); err != nil {
+			return err
+		}
+	}
+}
+
+func (n *nodeAddressController) update(txn statedb.WriteTxn, existing, new sets.Set[NodeAddress], reporter cell.HealthReporter) {
+	updated := false
+
+	// Insert new addresses that did not exist.
+	for addr := range new {
+		if !existing.Has(addr) {
+			updated = true
+			n.NodeAddresses.Insert(txn, addr)
+		}
+	}
+
+	// Remove addresses that were not part of the new set.
+	for addr := range existing {
+		if !new.Has(addr) {
+			updated = true
+			n.NodeAddresses.Delete(txn, addr)
+		}
+	}
+
+	if updated {
+		addrs := showAddresses(new)
+		n.Log.WithField("node-addresses", addrs).Info("Node addresses updated")
+		if reporter != nil {
+			reporter.OK(addrs)
+		}
+	}
+}
+
+func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.Set[NodeAddress]) {
+	addrs = sets.New[NodeAddress]()
+
+	if dev.Flags&net.FlagUp == 0 {
+		return
+	}
+
+	// Skip obviously uninteresting devices.
+	// We include the HostDevice as its IP addresses are consider node addresses
+	// and added to e.g. ipcache as HOST_IDs.
+	if dev.Name != defaults.HostDevice {
+		skip := false
+		for _, prefix := range defaults.ExcludedDevicePrefixes {
+			if strings.HasPrefix(dev.Name, prefix) {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			return
+		}
+	}
+
+	// ipv4 and ipv6 are set to true when the primary address is picked.
+	// Used to implement 'NodePort' and 'Primary' flags.
+	ipv4, ipv6 := false, false
+
+	for _, addr := range sortAddresses(dev.Addrs) {
+		// We keep the scope-based address filtering as was introduced
+		// in 080857bdedca67d58ec39f8f96c5f38b22f6dc0b.
+		if addr.Scope > uint8(n.AddressScopeMax) {
+			// Addresses are sorted by scope, so can stop early.
+			break
+		}
+		if addr.Addr.IsLoopback() {
+			continue
+		}
+
+		// Figure out if the address is usable for NodePort.
+		nodePort := false
+		primary := false
+		if dev.Selected && len(n.Config.NodePortAddresses) == 0 {
+			// The user has not specified IP ranges to filter on IPs on which to serve NodePort.
+			// Thus the default behavior is to use the primary IPv4 and IPv6 addresses of each
+			// device.
+			if addr.Addr.Is4() && !ipv4 {
```

Comment:
nit: had to double back and re-read the comment over `ipv4`, as it reads somewhat confusingly. Maybe `ipv4Found`/`ipv4Picked` instead?

Reply:
I struggled finding good names/pattern here and not too happy how it looks right now, but didn't figure out a better way. `ipv4Found` is definitely an approvement so going with that.